### ISSUE TITLE
[important] _compute_outlier_scores loops sample-by-sample calling scipy integrate_box_1d with a per-sample .cpu() …

### ISSUE THREAD

**Issue by martinez-hub** [mit-ll-responsible-ai/equine#178]
**Severity:** important  
**Location:** `src/equine/equine_protonet.py:736`  
**Found by review lenses:** ml-semantics, performance

### What is wrong

CONFIRMS requester item (4). Lines 736-742 iterate `for i in range(len(predictions))`, and per sample: `int(torch.argmax(predictions[i,:]))`, `ood_dists[i].detach().cpu().numpy()` (a device sync per row when on GPU), a scipy `integrate_box_1d` call, and a scalar write `ood_scores[i] = ...` (host->device copy per row on GPU). Measured on CPU with 20k rows: 0.37 s in this loop, i.e. essentially 100% of the 0.36 s `predict()` wall time; a vectorized torch equivalent (`torch.special.ndtr` over the KDE's stored `dataset` and bandwidth, grouped by predicted class) took 6 ms with max abs diff 1.2e-7. At 1M rows this is ~18 s of pure Python/scipy overhead, and far worse on GPU due to 2M synchronizing transfers.

### Suggested fix

Vectorize per predicted class: `pc = predictions.argmax(1); for lbl, kde in self.outlier_score_kde.items(): m = pc == lbl; d = torch.as_tensor(kde.dataset[0], dtype=ood_dists.dtype, device=ood_dists.device); sig = math.sqrt(kde.covariance[0,0]); cdf = torch.special.ndtr((ood_dists[m,None] - d[None,:]) / sig).mean(1); ood_scores[m] = cdf.clamp(0,1)` (this equals `1 - integrate_box_1d(x, inf)`). Store `d`/`sig` per class at fit time instead of the scipy object.

### Verification

An independent, refutation-biased pass re-read the code (and ran it where noted) and could not disprove the finding:

> Could not refute. src/equine/equine_protonet.py:735-742 `_compute_outlier_scores` iterates `for i in range(len(predictions))` and per row does `int(torch.argmax(predictions[i, :]))`, `ood_dists[i].detach().cpu().numpy()`, a scipy `gaussian_kde.integrate_box_1d(x, np.inf)` call, and a scalar `ood_scores[i] = ...` write; there is no vectorized path or guard. Lines 714-717 (`_fit_outlier_scores`) store a scipy gaussian_kde per class with a `# TODO convert to torch func` comment. It is reachable on every call of the public `predict()` at line 806 (also confirmed the double embedding: line 798 `compute_embeddings(X)` plus line 801 `self.model(X)` -> forward calls `compute_embeddings` again at line 400). Reproduced with /tmp/eq-venv/bin/python (torch 2.9.1, scipy 1.13.0): 4 classes, 20,000 rows -> existing loop 0.569 s vs vectorized `torch.special.ndtr((ood[m,None]-kde.dataset[0][None,:])/sqrt(kde.covariance[0,0])).mean(1)` 0.027 s, max abs diff 1.79e-7, so the suggested fix is an exact equivalent. On GPU the per-row `.cpu()` and scalar tensor write each force a device sync/transfer (2N transfers). Outputs are numerically correct, so this is a performance defect only, but it dominates predict() wall time on the library's main inference path and has a trivial exact fix; keeping 'important' rather than downgrading further.

### Also reported as

- OOD scoring loops over samples one at a time calling scipy integrate_box_1d per sample
- Both models run the embedding network twice per predict() and never disable autograd at inference (`src/equine/equine_protonet.py:798`)

---
_Found by an automated adversarial code review of `main` at 74a0cb5 (2026-09-18): ten independent reviewer agents, deduplication, then a skeptic agent per finding that tried to refute it by reading and running the code. Filed by @martinez-hub. Finding #13 of 72 confirmed._
